### DISCUSSION TITLE
How to assign a raw json string to a structure value?

### DISCUSSION THREAD

**Discussion by mgalera** [disneystreaming/smithy4s#954] [Q&A] [answered]
Hi everyone.

I'm trying to define a structure which have a field that contains a json object. I don't need it to be modeled with smithy itself, just to hold whatever value I set to it.

So far I'm trying to achieve this using a Document type :

`structure Stuff {
  @required
  field: Document
}`

Now, in my scala code I have a circe Json object that I want to assign to the field but I cannot find a way to create a Document from it (or from a json string).

So, my questions are
1. I'm right using the Document type here or there is another way to assign a raw json string to a structure field?
2. If Document is right, there is a straight way to create it from a Json value?

Thanks for the help!

**Comment by kubukoz** [comment, marked as answer]
let me see if I got it right: given the following

```scala
Stuff(Json.obj("hello" := 42))
```

you want the Json to be serialized as a string?

```json
{"field": "{\"hello\": 42}"}
```

In this case, your Smithy spec would have to have `structure Stuff { @required field: String }`, and you'd pass it as `Stuff(jsonobj.noSpaces /* or some other rendering */)`.

or do you just want to pass the `Json` value verbatim and let it be encoded as json?

```json
{"field": {"hello": 42}}
```

In this case, `Document` is fine, but you'll need to go from `Json` to `Document`. You can implement this on your side:

```scala
object JsonToDocumentFolder extends Folder[Document]:
  override def onNull: Document                    = Document.nullDoc
  override def onString(value: String): Document   = Document.fromString(value)
  override def onBoolean(value: Boolean): Document = Document.fromBoolean(value)
  override def onNumber(value: JsonNumber): Document = value.toInt
    .map(Document.fromInt)
    .orElse(
      value.toLong.map(Document.fromLong)
    )
    .getOrElse(
      Document.fromDouble(value.toDouble)
    )
  override def onArray(value: Vector[Json]): Document =
    Document.array(value.map(_.foldWith(this)))
  override def onObject(value: JsonObject): Document = Document.DObject(
    value.toMap.map { case (k, v) =>
      k -> v.foldWith(this)
    }
  )

val json: Json = ???

Stuff(json.foldWith(jsonToDocumentFolder))
```

**Reply by mgalera** [reply]
Hi Jakub.

The Folder worked like a charm, thanks you very much!

To make the reverse operation (Document to Json) I managed to use a Document.Decoder[Json]:

```
val decoder: Document.Decoder[Json] = new Document.Decoder[Json] {
      def decode(document: Document): Either[PayloadError, Json] = {
        def toJson(d: Document): Json = {
          d match {
            case DNumber(value)  => Json.fromBigDecimal(value)
            case DBoolean(value) => Json.fromBoolean(value)
            case DString(value)  => Json.fromString(value)
            case DNull           => Json.Null
            case DArray(value)   => Json.fromValues(value.map(toJson))
            case DObject(value) =>
              val newMap = value.map { case (k, v) => k -> toJson(v) }
              Json.fromJsonObject(JsonObject.fromMap(newMap))
          }
        }
        Right(toJson(document))
      }
    }

val json: Either[PayloadError, Json] = decoder.decode(smithyDocument)
```

I still have to deal with the PayloadError, but it seems like the right direction. Thanks again!
